src/operator/tensor/cast_storage-inl.cuh

```diff
@@ -9,15 +9,377 @@
 #include <mxnet/base.h>
 #include <mxnet/operator.h>
 
+#include <cub/cub.cuh>
+
 namespace mxnet {
 namespace op {
+using mshadow::cuda::kBaseThreadNum;
 
-inline void CastStorageDnsRspImpl(mshadow::Stream<gpu>* s, const TBlob& dns, NDArray* rsp) {
+inline void CastStorageDnsRspImpl(const OpContext& ctx, const gpu& gpu_dev, const TBlob& dns, NDArray* rsp) {
   LOG(FATAL) << "CastStorageDnsRspImpl gpu version is not implemented.";
 }
 
-inline void CastStorageDnsCsrImpl(mshadow::Stream<gpu>* s, const TBlob& dns, NDArray* csr) {
-  LOG(FATAL) << "CastStorageDnsCsrImpl gpu version is not implemented.";
+/*!
+ * \brief Thread kernel for initializing the indptr in a csr tensor.
+ * Parallelized by matrix rows: 1 thread/row
+ */
+struct FillCsrIndPtrThreadKernel {
```

Comment:
@eric-haibin-lin Yes, this kernel is essentially the same as FillCsrIndPtr kernel in .h file. However, since the function in .h cannot see this kernel if compiled with `g++`, I think we may have to keep them both in separate files.